File: backend/app/api/v1/me.py
from typing import Optional
from fastapi import APIRouter, Depends, Query, Response
from app.core.deps import require_nasabah
from app.schemas.nasabah import NasabahUpdateRequest
from app.services.dashboard_service import dashboard_service
from app.services.nasabah_service import nasabah_service
from app.services.receipt_service import receipt_service
from app.services.transaction_service import transaction_service
from app.utils.responses import success_response

router = APIRouter(prefix="/me", tags=["Nasabah Login Portal"])


@router.get("/nasabah")
def get_my_profile(current_user: dict = Depends(require_nasabah)):
    """Profil nasabah yang sedang login."""
    nasabah = current_user["nasabah"]
    nasabah["balance"] = nasabah_service.get_nasabah_balance(nasabah["id"])
    return success_response(data=nasabah)


@router.put("/nasabah")
def update_my_profile(
    data: NasabahUpdateRequest,
    current_user: dict = Depends(require_nasabah)
):
    """Ubah data profil nasabah yang sedang login."""
    nasabah_id = current_user["nasabah_id"]
    nasabah = nasabah_service.update_nasabah(
        nasabah_id, data, updater_id=current_user["id"]
    )
    nasabah["balance"] = nasabah_service.get_nasabah_balance(nasabah_id)
    return success_response(
        data=nasabah,
        message="Profil data diri Anda berhasil diperbarui."
    )


@router.get("/balance")
def get_my_balance(current_user: dict = Depends(require_nasabah)):
    """Saldo tabungan nasabah yang sedang login."""
    nasabah_id = current_user["nasabah_id"]
    balance = nasabah_service.get_nasabah_balance(nasabah_id)
    return success_response(
        data={
            "nasabah_id": nasabah_id,
            "customer_id": current_user["nasabah"]["customer_id"],
            "account_no": current_user["nasabah"]["account_no"],
            "name": current_user["nasabah"]["name"],
            "balance": balance,
        }
    )


@router.get("/dashboard")
def get_my_dashboard(current_user: dict = Depends(require_nasabah)):
    """Ringkasan dashboard untuk nasabah yang sedang login."""
    data = dashboard_service.get_nasabah_dashboard(current_user["nasabah_id"])
    return success_response(data=data)


@router.get("/transactions")
def get_my_transactions(
    start_date: Optional[str] = Query(None, description="Tanggal awal (YYYY-MM-DD)"),
    end_date: Optional[str] = Query(None, description="Tanggal akhir (YYYY-MM-DD)"),
    type: Optional[str] = Query(None, description="SETOR atau TARIK"),
    category_id: Optional[str] = Query(None, description="ID Kategori"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    current_user: dict = Depends(require_nasabah)
):
    """Daftar riwayat transaksi milik nasabah yang sedang login."""
    items, total_items = transaction_service.list_transactions(
        nasabah_id=current_user["nasabah_id"],
        start_date=start_date,
        end_date=end_date,
        type_filter=type,
        category_id=category_id,
        page=page,
        page_size=page_size
    )
    total_pages = (total_items + page_size - 1) // page_size if total_items > 0 else 0
    return success_response(
        data={
            "items": items,
            "pagination": {
                "page": page,
                "page_size": page_size,
                "total_items": total_items,
                "total_pages": total_pages,
            }
        }
    )


@router.get("/transactions/{transaction_id}/receipt")
def get_my_receipt(
    transaction_id: str,
    format: Optional[str] = Query("json", description="json atau pdf"),
    current_user: dict = Depends(require_nasabah)
):
    """Bukti transaksi milik nasabah yang sedang login (JSON atau Download PDF)."""
    if format == "pdf":
        pdf_bytes = receipt_service.generate_receipt_pdf(transaction_id, current_user)
        filename = f"bukti-transaksi-{transaction_id[:8]}.pdf"
        return Response(
            content=pdf_bytes,
            media_type="application/pdf",
            headers={"Content-Disposition": f"inline; filename={filename}"}
        )
    
    receipt_data = receipt_service.get_receipt_data(transaction_id, current_user)
    return success_response(data=receipt_data)

File: backend/src/routes/me.ts
import { Hono } from 'hono';
import { AppEnv } from '../types/index.js';
import { authMiddleware } from '../middleware/auth.js';
import { requireNasabah } from '../middleware/roleGuard.js';
import { nasabahService } from '../services/nasabahService.js';
import { dashboardService } from '../services/dashboardService.js';
import { transactionService } from '../services/transactionService.js';
import { receiptService } from '../services/receiptService.js';
import { successResponse } from '../utils/response.js';
import { validateJson } from '../utils/validator.js';
import { nasabahUpdateSchema } from '../schemas/nasabah.schema.js';

export const meRouter = new Hono<AppEnv>();

// All routes require NASABAH
meRouter.use('*', authMiddleware, requireNasabah);

// GET /api/v1/me/nasabah
meRouter.get('/nasabah', async (c) => {
  const user = c.get('user');
  const nasabah = await nasabahService.getNasabahById(user.nasabah_id || user.id);
  return successResponse(c, nasabah);
});

// PUT /api/v1/me/nasabah
meRouter.put('/nasabah', validateJson(nasabahUpdateSchema), async (c) => {
  const user = c.get('user');
  const data = c.req.valid('json' as any);
  const nasabah = await nasabahService.updateNasabah(user.nasabah_id || user.id, data, user.id);
  return successResponse(c, nasabah, 'Profil data diri Anda berhasil diperbarui.');
});

// GET /api/v1/me/balance
meRouter.get('/balance', async (c) => {
  const user = c.get('user');
  const balance = await nasabahService.getNasabahBalance(user.nasabah_id || user.id);
  return successResponse(c, {
    nasabah_id: user.nasabah_id,
    customer_id: user.nasabah?.customer_id || '',
    account_no: user.nasabah?.account_no || '',
    name: user.nasabah?.name || '',
    balance,
  });
});

// GET /api/v1/me/dashboard
meRouter.get('/dashboard', async (c) => {
  const user = c.get('user');
  const data = await dashboardService.getNasabahDashboard(user.nasabah_id || user.id);
  return successResponse(c, data);
});

// GET /api/v1/me/transactions
meRouter.get('/transactions', async (c) => {
  const user = c.get('user');
  const query = c.req.query();
  const result = await transactionService.listTransactions({
    nasabah_id: user.nasabah_id || user.id,
    start_date: query.start_date,
    end_date: query.end_date,
    type: query.type,
    category_id: query.category_id,
    page: query.page ? parseInt(query.page, 10) : 1,
    page_size: query.page_size ? parseInt(query.page_size, 10) : 20,
  });
  return successResponse(c, result);
});

// GET /api/v1/me/transactions/:id/receipt
meRouter.get('/transactions/:id/receipt', async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const format = c.req.query('format');

  if (format === 'pdf') {
    const pdfBuffer = await receiptService.generateReceiptPdf(id, user);
    c.header('Content-Type', 'application/pdf');
    c.header('Content-Disposition', `inline; filename="bukti-transaksi-${id.slice(0, 8)}.pdf"`);
    return c.body(new Uint8Array(pdfBuffer));
  }

  const receiptData = await receiptService.getReceiptData(id, user);
  return successResponse(c, receiptData);
});
